BJS2-40421 Оплата через DMS - 2

kripersh27 opened the pull request:
Клиент инициирует платеж, отправляет запрос ко мне в пэймент, проверяю идемпотентность(это хэш из данных в дто), присваиваю статус Пендинг и статус баланса - "не верефицирован", сохраняю операцию и отправляю в аккаунт сервис через редис\кафку дто с данными и статусом операции - "Пендинг", аккаунт сервис сверяет баланс и отправляет через кафку\редис инфу о том, достаточен баланс для совершения данной операции, если баланса не хватает, то операции присваивается статус баланса - "не хватает" и статус самой операции еррор, обновляются данные в бд и об этом сообщаем клиенту, если денег на балансе достаточно, то ставим статус баланса - "достаточно средств", и статус операции - "авторизовано", далее у этих операций есть время, когда они должны быть автоматически клирингованы, если до этого времени клиент отменит операцию своим запросом, то происходит апдейт операции на статус "Канцелятион" и отправляется инфа в аккаунт сервис о том, что запрос отменен и статус операции "Канцелятион", далее аккаунт сервис это подтверждает и отправляет сообщение успех\уошибка, если же клиент подтвердил операцию в ручном режиме или вышло время для автоматического подтверждения, то снова происходит апдейт данных, присваивается статус - клиринг, отправляется в редис\кафку инфа со статусом операции "Клиринг", и если же в процессе клиринга произошла какая-то ошибка, то сущность снова апдейтится со статусом Фэйлед, и уже отправляются данные с типом операции Эррор. Аккаунт сервис на каждом этапе подтверждает информацию, что либо всё хорошо, либо ошибка, про ошибку мы сообщаем клиенту, если всё хорошо, то операции присваивается статус операции с "FINISHED_". Клиент так же может проверять статус операции через эндпоинт.

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Добавлена сущность PendingOperation, а так же контроллер, сервисный слой, редис конфиг, публешер редис, тесты

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Разделение ручного и автоматического подтверждения операции

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Поправил тесты

sultan-inked commented on src/main/java/faang/school/paymentservice/validator/PendingOperationValidator.java:
```diff
@@ -0,0 +1,44 @@
+        }
+        if (operation.getClearScheduledAt() == null || operation.getClearScheduledAt().isBefore(LocalDateTime.now())) {
+            throw new InvalidOperationException("Clear scheduled time must be in the future");
+        }
```
Разве не проще аннотации на дто повесить? Возможно чего то не понял

sultan-inked commented on src/main/java/faang/school/paymentservice/repository/PendingOperationRepository.java:
```diff
@@ -0,0 +1,18 @@
+public interface PendingOperationRepository extends JpaRepository<PendingOperation, UUID> {
+    Optional<PendingOperation> findByIdAndStatus(UUID id, OperationStatus status);
+    List<PendingOperation> findByStatusAndClearScheduledAtBefore(OperationStatus status, LocalDateTime dateTime);
+    Optional<PendingOperation> findByIdempotencyKey(String idempotencyKey);
```
Методы возможно все таки стоит разделять строчками, сложно читать

cemeh666 replied:
++

sultan-inked commented on src/main/java/faang/school/paymentservice/service/PendingOperationService.java:
```diff
@@ -0,0 +1,88 @@
+    public UUID initiateOperation(PendingOperation operation) {
+        pendingOperationValidator.validateInitiateOperation(operation);
+
+        Optional<PendingOperation> existingOperation = pendingOperationRepository.findByIdempotencyKey(operation.getIdempotencyKey());
```
Длинная строчка

sultan-inked commented on src/main/java/faang/school/paymentservice/service/PendingOperationService.java:
```diff
@@ -0,0 +1,88 @@
+
+    @SendPendingOperationMessage(OperationType.CLEARING)
+    @Transactional
+    public void confirmOperation(UUID operationId, boolean isManual) {
```
Я бы сделал два метода, имхо

kripersh27 replied:
Можно, но лучше чтобы тут логика была, так понятнее

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Добавлена функциональность отправки сообщения через кафку, переключить можно через профиль "кафка"

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Добавлен гайд как запустить кафку

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Добавлен второй бин для редис темплейт с джексон

cemeh666 commented on src/main/java/faang/school/paymentservice/config/RedisConfig.java:
```diff
@@ -23,6 +24,18 @@ public JedisConnectionFactory jedisConnectionFactory() {
         return new JedisConnectionFactory(config);
     }
 
+    @Bean("redisTemplatewithJackson")
```
redisTemplateWithJackson
глаз резануло)

kripersh27 replied:
А как по другому?)

cemeh666 replied:
у тебя **redisTemplatewithJackson**
а я прошу так **redisTemplateWithJackson** сделать

kripersh27 replied:
А, понял, не заметил

cemeh666 commented on src/main/java/faang/school/paymentservice/controller/PendingOperationController.java:
```diff
@@ -0,0 +1,42 @@
+@RestController
+public class PendingOperationController {
+    private final PendingOperationService pendingOperationService;
+    final PendingOperationMapper pendingOperationMapper;
```
private забыл видимо

cemeh666 commented on src/main/java/faang/school/paymentservice/controller/PendingOperationController.java:
```diff
@@ -0,0 +1,42 @@
+    @PostMapping("/confirm/{id}")
+    public ResponseEntity<Void> confirmOperation(@PathVariable UUID id) {
+        pendingOperationService.confirmOperation(id, true);
+        return ResponseEntity.ok().build();
```
для пустого ответа лучше использовать 204 статус. 
ResponseEntity.ok() вернёт 200
@ResponseStatus(HttpStatus.NO_CONTENT)

cemeh666 commented on src/main/java/faang/school/paymentservice/exception/OperationNotFoundException.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.http.HttpStatus;
+
+public class OperationNotFoundException extends ApiException {
+    public OperationNotFoundException(String message) {
```
message можно константой сделать, чтобы в конструкторе каждый раз не писать

cemeh666 commented on src/main/java/faang/school/paymentservice/mapper/PendingOperationMapper.java:
```diff
@@ -0,0 +1,27 @@
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface PendingOperationMapper {
+
```
лишняя строка

cemeh666 commented on src/main/resources/db/changelog/changeset/payment_V001__create_pending_operation.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE pending_operation (
+                                   id UUID PRIMARY KEY,
+                                   account_id UUID NOT NULL,
+                                   idempotency_key VARCHAR(255) NOT NULL UNIQUE,
+                                   amount bigint NOT NULL,
+                                   currency VARCHAR(3) NOT NULL,
+                                   status VARCHAR(20) NOT NULL,
+                                   clear_scheduled_at timestamptz NOT NULL,
+                                   created_at timestamptz NOT NULL DEFAULT current_timestamp,
+                                   updated_at timestamptz DEFAULT current_timestamp
+);
```
Оооочень большой отступ

cemeh666 commented on src/main/resources/readmeKafka.md:
````diff
@@ -0,0 +1,42 @@
+```shell
+docker-compose up -d
+```
+Всё, контейнеры запущены, можно пулить ветку и тестить. 
````
у кафки есть какой веб интерефейс, чтобы смотреть, что в ней?
если есть укажи как в него зайти и куда смотреть

kripersh27 replied:
Каф дроп для этого как раз, зайти по текущим настройкам на localhost:9101 в браузере

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Изменения после кросс ревью

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Изменения после кросс ревью

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Добавлена логика проверки баланса у аккаунта в аккаунт сервисе

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Поправил публишер, добавил обджект маппер

sultan-inked committed:
BJS2-40421 add user context

20cognosce commented on src/main/java/faang/school/paymentservice/config/KafkaConfig.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Map;
+
+@Configuration
+@Profile("kafka")
```
Что значит этот профиль? Ты в пропертях конфигурацию просто прописал, какой профиль

20cognosce replied:
@ConfigurationProperties
@ConditionalOnProperty(kafka/redis)

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -3,6 +3,43 @@ spring:
+      channel:
+        pending_operation: pending_operation
+
+  kafka:
```
? 
И как другие будут запускать сервис без кафки

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -3,6 +3,43 @@ spring:
+    topic:
+      pending_operation: pending_operation
+  profiles:
+    active: redis
```
Куда ты спрятал...

20cognosce replied:
kafka - это не профиль
redis - это не профиль

это конфигурации

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -3,6 +3,43 @@ spring:
+        pending_operation: pending_operation
+
+  kafka:
+    bootstrap-servers: host.docker.internal:9092
```
Через докер всё равно делал

kripersh27 replied:
Я локально у себя его поднял, в инфру не лазил, у других сломаться ничего не должно

20cognosce commented on src/test/java/faang/school/paymentservice/service/OperationMessageServiceTest.java:
```diff
@@ -0,0 +1,69 @@
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class OperationMessageServiceTest {
```
Интеграционные тесты почему не завезли?

20cognosce commented on src/main/java/faang/school/paymentservice/config/KafkaConfig.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Bean
+    public KafkaTemplate<String, Object> kafkaTemplate() {
+        return new KafkaTemplate<>(producerFactory());
```
Зачем ты второй инстанс создаешь, у тебя уже есть бин

20cognosce commented on src/main/java/faang/school/paymentservice/config/KafkaConfig.java:
```diff
@@ -0,0 +1,52 @@
+        configs.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+        return new KafkaAdmin(configs);
+    }
+    public NewTopic createNewTopic(String topicName, int partitions, short replicas) {
+        return TopicBuilder.name(topicName)
+                .partitions(partitions)
+                .replicas(replicas)
+                .build();
+    }
```
?

kripersh27 replied:
Это общий метод для создания топиков, чтобы его реализация была уже непосредственно в паблишере, вдруг еще какой паблишер сделать нужно будет

20cognosce commented on src/main/java/faang/school/paymentservice/config/RedisConfig.java:
```diff
@@ -23,6 +24,18 @@ public JedisConnectionFactory jedisConnectionFactory() {
         return new JedisConnectionFactory(config);
     }
 
+    @Bean("redisTemplateWithJackson")
```
Может дополнить уже имеющийся RedisTemplate? А не второй бин создавать ради сериализатора

kripersh27 replied:
template.setValueSerializer(new GenericJackson2JsonRedisSerializer());
С дженериком у Султана не работает, но его использует Евгений в своем коде, поэтому чтобы не ломать код, и чтобы работало у Султана, пришлось так сделать

20cognosce commented on src/main/java/faang/school/paymentservice/publisher/kafka/KafkaEventPublisher.java:
```diff
@@ -0,0 +1,62 @@
+        this.kafkaAdmin = kafkaAdmin;
+        this.kafkaConfig = kafkaConfig;
+        this.objectMapper = objectMapper;
+        createTopicIfNotExists();
+    }
+
+    private void createTopicIfNotExists() {
+        NewTopic newTopic = kafkaConfig.createNewTopic(topic, 1, (short) 1);
+
+        try {
+            kafkaAdmin.createOrModifyTopics(newTopic);
+            log.info("Kafka topic '{}' checked or created successfully", topic);
+        } catch (Exception e) {
+            log.error("Failed to create topic '{}': {}", topic, e.getMessage());
+        }
+    }
```
Ты таблицы в базе данных тоже через джава код создаешь?
PostgresTable.createNewColumn

Можно разделить настройку кафки (не через приложение, а либо руками, либо отдельными скриптами) и подключение к ней, это не правильно

20cognosce commented on src/main/resources/db/changelog/changeset/payment_V001__create_pending_operation.sql:
```diff
@@ -0,0 +1,14 @@
+    updated_at timestamptz DEFAULT current_timestamp
+);
+
+CREATE INDEX index_pending_operation_status ON pending_operation (status);
+CREATE INDEX index_pending_operation_status_clear_scheduled_at ON pending_operation (status, clear_scheduled_at);
```
На что ты индексы создал...
На статус и на дату? Зачем?

kripersh27 replied:
У меня есть джоба, которая периодически проверяет в бд по этим полям операции, и когда время подошло, меняет их статус, поэтому я тут решил, что если это частый запрос, то лучше сделать составной индекс по этим двум полям.

kripersh27 replied:
В задании сказано - подумайте куда лучше создать индексы

20cognosce replied:
Там также написано "если они необходимы"
Конкретно эти два индекса бесполезны или даже вредны, потому что поля по которым они создаются обладают слабой уникальностью

Такими индексами **условно** ты создал хэшмапу в которой два ключа, и каждому ключу соответствует значение из списка, в котором тысяча элементов

20cognosce commented on src/main/resources/db/changelog/changeset/payment_V001__create_pending_operation.sql:
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE pending_operation (
```
Почему таблица называется pending_iperation, pending - это и есть статус?

kripersh27 replied:
Переводится как "ожидающая выполнения операция"

20cognosce commented on src/main/resources/db/changelog/changeset/payment_V001__create_pending_operation.sql:
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE pending_operation (
+    id UUID PRIMARY KEY,
+    account_id UUID NOT NULL,
+    idempotency_key VARCHAR(255) NOT NULL UNIQUE,
+    amount bigint NOT NULL,
+    currency VARCHAR(3) NOT NULL,
+    status VARCHAR(20) NOT NULL,
+    clear_scheduled_at timestamptz NOT NULL,
+    created_at timestamptz NOT NULL DEFAULT current_timestamp,
+    updated_at timestamptz DEFAULT current_timestamp
```
Форматирование столбцов

20cognosce commented on src/main/java/faang/school/paymentservice/validator/PendingOperationValidator.java:
```diff
@@ -0,0 +1,53 @@
+    public void validateBalance(UUID accountId, BigDecimal amount) {
+        BalanceResponseDto balance = accountServiceClient.findBalanceByAccountId(accountId);
+
+        BigDecimal availableBalance = balance.currentBalance().subtract(balance.authBalance());
+        if (availableBalance.compareTo(amount) < 0) {
+            throw new InsufficientBalanceException("Insufficient balance for account ID: " + accountId);
+        }
```
Ты не можешь это делать, нет, ты за это не отвечаешь

Султан рисовал архитектуру задачи, что сейчас фактически реализовано соответствует?

20cognosce replied:
Что это, господа, что это за дизайн...

20cognosce replied:
Это вообще не проверка, это шляпа, с балансом там может быть что угодно, ты тут никакие блокировки не предпринимаешь и знать не можешь об актуальном состоянии баланса.

10 юзеров запросили с одного баланса по тысяче рублей, уже ими покупки оплатили, уже сами себе клиринг подтвердили. И потом по редису до аккаунт_сервиса доходит 10 платежей, а денег зарезервировал он только для одного, остальные 9 ушли в ошибку, а нам тут пофиг ваще

20cognosce commented on src/main/java/faang/school/paymentservice/mapper/PendingOperationMapper.java:
```diff
@@ -0,0 +1,26 @@
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface PendingOperationMapper {
+    @Mapping(target = "id", expression = "java(java.util.UUID.randomUUID())")
```
Ты не можешь так генерировать айдишник, это должен делать JPA

20cognosce commented on src/main/java/faang/school/paymentservice/service/PendingOperationService.java:
```diff
@@ -0,0 +1,82 @@
+    @SendPendingOperationMessage(OperationType.AUTHORIZATION)
+    @Transactional
+    public UUID initiateOperation(PendingOperation operation) {
+        pendingOperationValidator.validateIdempotencyKey(operation.getIdempotencyKey());
```
Где он задается? С фронта приходит?

kripersh27 replied:
Да, с фронта приходит, если я буду создавать, то идемпотентность не соблюсти

20cognosce replied:
Может все-таки соблюсти, если ключ идемпотентности генерировать по атрибутам?
От кого, кому, тип операции, сумма
Они ведь статичны

И условно за последнюю минуту генерировать по всем этим полям хэш-код, который будет и ключом идемпотентности

Если отправлять сколько угодно запросов с такими же параметрами в течение минуты, то вторая операция не пройдет

Подумай над такой фичей

20cognosce commented on src/main/java/faang/school/paymentservice/service/PendingOperationService.java:
```diff
@@ -0,0 +1,82 @@
+    @Transactional
+    public UUID initiateOperation(PendingOperation operation) {
+        pendingOperationValidator.validateIdempotencyKey(operation.getIdempotencyKey());
+        pendingOperationValidator.validateBalance(operation.getAccountId(), operation.getAmount());
+        pendingOperationRepository.save(operation);
+        log.info("Operation initiated with ID: {}", operation.getId());
```
Баланс ты не валидируешь, удалить
account_service сам решает, достаточен баланс или нет, ты только запрашиваешь

save метод сам генерирует id, и возвращает сущность, с сохранённым в базе id

20cognosce commented on src/main/java/faang/school/paymentservice/service/PendingOperationService.java:
```diff
@@ -0,0 +1,82 @@
+    public UUID initiateOperation(PendingOperation operation) {
+        pendingOperationValidator.validateIdempotencyKey(operation.getIdempotencyKey());
+        pendingOperationValidator.validateBalance(operation.getAccountId(), operation.getAmount());
+        pendingOperationRepository.save(operation);
```
Есть только операция.
PENDING
CANCELLED
CLEARED
Это всё её статусы

20cognosce commented on src/main/java/faang/school/paymentservice/aop/aspects/PendingOperationMessageAspect.java:
```diff
@@ -0,0 +1,52 @@
+
+    @AfterReturning(pointcut = "@annotation(sendPendingOperationMessage)", returning = "result")
+    public void afterReturning(JoinPoint joinPoint, SendPendingOperationMessage sendPendingOperationMessage, Object result) {
+        OperationType operationType = sendPendingOperationMessage.value();
+        UUID operationId = null;
+
+        for (Object arg : joinPoint.getArgs()) {
+            if (arg instanceof UUID) {
+                operationId = (UUID) arg;
+                break;
+            } else if (arg instanceof PendingOperation) {
+                operationId = ((PendingOperation) arg).getId();
+                break;
+            }
+        }
+
+        if (operationId == null && result instanceof UUID) {
+            operationId = (UUID) result;
+        } else if (operationId == null && result instanceof PendingOperation) {
+            operationId = ((PendingOperation) result).getId();
+        }
+
+        if (operationId == null) {
+            log.error("Failed to find operationId in method arguments or return value");
+            throw new IllegalArgumentException("Failed to find operationId in method arguments or return value");
+        }
+
+        log.info("Publishing message with operation ID: {} and operation type: {}", operationId, operationType);
```
Какой это ужас...
Один раз плохому научил и теперь везде у нас аспекты будут

Отправка события в account_service - **неотъемлемая** часть метода твоего сервиса, не нужны тут аспекты, удали всё это

20cognosce commented on src/main/java/faang/school/paymentservice/service/OperationMessageService.java:
```diff
@@ -0,0 +1,45 @@
+    private final EventPublisher<OperationMessage> eventPublisher;
+    private final PendingOperationRepository pendingOperationRepository;
+
+    @Autowired
```
Зачем ты тут написал autowired, он ничего не делает

Конструктор через ломбок можешь сгенерировать

20cognosce commented on src/main/java/faang/school/paymentservice/service/OperationMessageService.java:
```diff
@@ -0,0 +1,45 @@
+    public void sendOperationMessage(UUID operationId, OperationType operationType) {
+        log.debug("Sending operation message for operationId: {}, operationType: {}", operationId, operationType);
+
+        PendingOperation pendingOperation = pendingOperationRepository.findById(operationId)
+                .orElseThrow(() -> new RuntimeException("Operation not found with ID: " + operationId));
```
А можно не id, а сразу сущность
Из метода сервиса сразу вызывай после сохранения в БД через saveAndFlush

20cognosce replied:
Можешь асинхронно вызывать, @Async хорошо ложится на этот метод

20cognosce commented on src/main/java/faang/school/paymentservice/repository/PendingOperationRepository.java:
```diff
@@ -0,0 +1,20 @@
+
+@Repository
+public interface PendingOperationRepository extends JpaRepository<PendingOperation, UUID> {
+    Optional<PendingOperation> findByIdAndStatus(UUID id, OperationStatus status);
```
Не понимаю, чем поиск по id будет отличаться от поиска по id и статусу

20cognosce commented on src/main/java/faang/school/paymentservice/service/PendingOperationService.java:
```diff
@@ -0,0 +1,82 @@
+
+    private void updateOperationStatus(PendingOperation operation, OperationStatus status) {
+        operation.setStatus(status);
+        operation.setUpdatedAt(LocalDateTime.now());
```
Можно аннотациями апдейтить таймстемпы

20cognosce commented on src/main/java/faang/school/paymentservice/service/PendingOperationService.java:
```diff
@@ -0,0 +1,82 @@
+    @Transactional
+    public void confirmOperation(UUID operationId, boolean isManual) {
+        PendingOperation operation = getOperationForProcessing(operationId, OperationStatus.PENDING);
+        try {
+            if (isManual) {
+                pendingOperationValidator.validateManualConfirmation(operation);
+            } else {
+                pendingOperationValidator.validateAutomaticConfirmation(operation);
```
Так
Я не понял

С каких пор мы инициализируем клиринг без какого либо ответа от аккаунт_сервиса? Тебе аккаунт_сервис должен ответить, ЧТО ЗАРЕЗЕРВИРОВАЛ СУММУ НА ЭТАПЕ АВТОРИЗАЦИИ И ГОТОВ К СПИСАНИЮ

Создаем - списываем - создаем - списываем, никого не спрашиваем, как красиво...

Вы с Султаном прорабатывали архитектуру?

20cognosce commented on src/main/java/faang/school/paymentservice/controller/PendingOperationController.java:
```diff
@@ -0,0 +1,42 @@
+    private final PendingOperationMapper pendingOperationMapper;
+
+    @PostMapping("/initiate")
+    public ResponseEntity<UUID> initiateOperation(@RequestBody PendingOperationDto operationDto) {
```
Объясни как ты реализовал идемпотентность

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Добавил листенера с возможностью проверки баласа аккаунта

chudoudoribakit committed:
Merge remote-tracking branch 'origin/feature-BJS2-40421' into feature-BJS2-40421

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Правки и интеграционный тест

kripersh27 committed:
Merge branch 'chimera-master-stream6' into feature-BJS2-40421

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Правки, убрал создание топиков у кафки в докер компоуз файл, добавил эндпоинт для просмотра клиентом статуса операции

chudoudoribakit committed:
Merge remote-tracking branch 'origin/feature-BJS2-40421' into feature-BJS2-40421

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Правки

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Правки тестов

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Правки

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -3,6 +3,53 @@ spring:
+      database-platform: org.hibernate.dialect.PostgreSQLDialect
+      hibernate:
+        ddl-auto: none
+      show-sql: true
+      properties:
+        hibernate:
+          format_sql: true
```
Вырубайте sql логи, они для дебага

20cognosce commented on src/main/java/faang/school/paymentservice/service/ClearingJob.java:
```diff
@@ -0,0 +1,38 @@
+    private final OperationMessageService operationMessageService;
+
+    @Async
+    @Scheduled(fixedDelayString = "${app.clearing.job.interval}")
```
Scheduled итак в отдельном потоке стартует

kripersh27 replied:
Выше ты мне сказал, что сюда отлично подойдет аннотация Асинк)

20cognosce commented on src/main/java/faang/school/paymentservice/service/ClearingJob.java:
```diff
@@ -0,0 +1,38 @@
+        LocalDateTime now = LocalDateTime.now();
+        List<PendingOperation> operationsForClearing = pendingOperationService.getOperationsForClearing(now);
+
+        operationsForClearing.forEach(operation -> {
```
Лучше тут не цикл, а асинхронные таски через экзекьютор сервис запускай

20cognosce commented on src/main/java/faang/school/paymentservice/dto/OperationMessage.java:
```diff
@@ -0,0 +1,26 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class OperationMessage {
```
Один топик всё-таки у вас и одно сообщение и на авторизацию и на клиринг
Ну окей, пойдет, но лучше разделить в реальных условиях, чтобы не путаться и не сортировать каждое сообщение по статусу в if else

20cognosce reviewed (changes requested):
А где прослушивание сообщений от Султана об успешном клиринге? Не увидел, должно быть, потому что и на этапе клиринга может оказаться ошибка на стороне account_service, которая заставит всю транзакцию откатить, а ты уже всё применил у себя

В целом супер гуд, машина 💪

Нюансы доделать

20cognosce commented:
Помимо этого, ребята, ждут от вас двоих ДЕМО к четвергу или воскресенью по этой задаче, она мощная и интересная, нужно показать вашу работу и функционал

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Правки после ревью

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Добавил тесты

chudoudoribakit committed:
BJS2-40421 Оплата через DMS - 2

Поправил генерацию идемпотентного ключа

20cognosce commented on src/main/java/faang/school/paymentservice/listener/kafka/AuthPaymentKafkaListener.java:
```diff
@@ -0,0 +1,36 @@
+    private final ObjectMapper objectMapper;
+    private final CheckingBalanceService checkingBalanceService;
+
+    @KafkaListener(topics ="${spring.kafka.topic.auth-payment-response}")
+    public void checkingAccountBalanceListener(String message) {
+        try {
+            CheckingPaymentStatusAndBalance event = objectMapper.readValue(message, CheckingPaymentStatusAndBalance.class);
+            checkingBalanceService.checkBalance(event, event.getStatus());
+
+        } catch (JsonProcessingException exception) {
+            log.error("Unexpected error, listen topic: {}", topic, exception);
+            throw new RuntimeException(exception);
+        }
+    }
```
Нельзя было всех объявить в одном классе?

20cognosce commented on src/main/java/faang/school/paymentservice/listener/redis/CancelPaymentResponseListener.java:
```diff
@@ -0,0 +1,39 @@
+@Slf4j
+@ConditionalOnProperty(prefix = "app", name = "messaging", havingValue = "redis")
+@Component
+public class CancelPaymentResponseListener extends AbstractEventListener<PaymentStatusResponseDto> {
+    private final CheckPaymentStatusService checkPaymentStatusService;
+
+    public CancelPaymentResponseListener(ObjectMapper objectMapper,
+                                         @Value("${spring.data.redis.channel.cancel-payment-response}") String topic,
+                                         CheckPaymentStatusService checkPaymentStatusService) {
+        super(objectMapper, new ChannelTopic(topic));
+        this.checkPaymentStatusService = checkPaymentStatusService;
+    }
+
+    @Override
+    public void saveEvent(PaymentStatusResponseDto event) {
+        checkPaymentStatusService.checkPaymentStatus(event);
+    }
+
+    @Override
+    public Class<PaymentStatusResponseDto> getEventType() {
+        return PaymentStatusResponseDto.class;
+    }
+
+    @Override
+    public void handleException(Exception exception) {
+        log.error("Error processing cancel-payment-response event", exception);
+    }
```
Аналогично, дублирование кода

20cognosce commented on src/main/java/faang/school/paymentservice/publisher/kafka/KafkaEventPublisher.java:
```diff
@@ -0,0 +1,56 @@
+                               @Value("${spring.kafka.topic.error-payment-request}") String errorChannel) {
+        this.kafkaTemplate = kafkaTemplate;
+        this.objectMapper = objectMapper;
+        this.topicMap = Map.of(
+                OperationStatus.PENDING, authTopic,
+                OperationStatus.CANCELLATION, cancelTopic,
+                OperationStatus.CLEARING, clearingTopic,
+                OperationStatus.ERROR, errorChannel
+        );
```
Ну не от статуса зависит топик, нет
От бизнес-процесса зависит топик

20cognosce commented on src/main/java/faang/school/paymentservice/service/CheckPaymentStatusService.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    private void handleErrorStatus(PendingOperation operation) {
+        switch (operation.getStatus()) {
+            case CANCELLATION, ERROR -> {
+                updateOperationStatus(operation, OperationStatus.ERROR);
+                throw new ErrorOperationException("Error occurred during the cancellation operation.");
+            }
+            case CLEARING -> {
+                updateOperationStatus(operation, OperationStatus.ERROR);
+                throw new ErrorOperationException("Error occurred during the clearing operation.");
+            }
+            default ->
+                    log.warn("Unhandled operation status {} for error event on operation {}",
+                            operation.getStatus(), operation.getId());
+        }
+    }
+
+    private void handleSuccessStatus(PendingOperation operation, PaymentStatusResponseDto event) {
+        switch (operation.getStatus()) {
+            case CANCELLATION -> {
+                updateOperationStatus(operation, OperationStatus.FINISHED_CANCELLATION);
+                log.info("Payment with id: {} was cancelled", event.getOperationId());
+            }
+            case ERROR -> {
+                updateOperationStatus(operation, OperationStatus.FINISHED_ERROR);
+                log.info("Payment with id: {} was aborted", event.getOperationId());
+            }
+            case CLEARING -> {
+                updateOperationStatus(operation, OperationStatus.FINISHED_CLEARING);
+                log.info("Payment with id: {} was successful", event.getOperationId());
+            }
+            default ->
+                    log.warn("Unexpected operation status {} for successful event on operation {}",
+                            operation.getStatus(), operation.getId());
+        }
+    }
```
А в чем разница?

20cognosce commented on src/main/java/faang/school/paymentservice/service/CheckingBalanceService.java:
```diff
@@ -0,0 +1,61 @@
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class CheckingBalanceService {
```
Нельзя ветвить логику приложения опираясь на статус операции, это неправильный подход
Статус переименовывается, и что дальше, в кучу методов лезть менять, и еще бы не перепутать 

Просто в зависимости от того, из какого топика приходит событие, нужно вызывать тот или иной код.
От топика, не от статуса.

20cognosce commented on src/main/java/faang/school/paymentservice/service/ClearingJob.java:
```diff
@@ -0,0 +1,42 @@
+        LocalDateTime now = LocalDateTime.now();
+        List<PendingOperation> operationsForClearing = pendingOperationService.getOperationsForClearing(now);
+
+        List<CompletableFuture<Void>> futures = operationsForClearing.stream()
```
Зачем тебе CompletableFuture
Почему не простой экзекьютор

Какой ответ ты ожидаешь для коллбека? Никакой

20cognosce commented on src/main/java/faang/school/paymentservice/util/IdempotencyKeyGenerator.java:
```diff
@@ -0,0 +1,34 @@
+                operation.getAmount().toString(),
+                operation.getCurrency(),
+                operation.getCategory(),
+                operation.getClearScheduledAt().toString()
```
До минут может делать truncate

20cognosce commented on src/main/java/faang/school/paymentservice/util/IdempotencyKeyGenerator.java:
```diff
@@ -0,0 +1,34 @@
+        try {
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
+            byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));
+            return Base64.getEncoder().encodeToString(hash);
```
Хорошо, только меня интересует длина такого хэша

20cognosce commented on src/main/resources/db/changelog/changeset/payment_V001__create_pending_operation.sql:
```diff
@@ -0,0 +1,15 @@
+    id                     UUID PRIMARY KEY,
+    source_account_id      UUID         NOT NULL,
+    target_account_id      UUID         NOT NULL,
+    idempotency_key        VARCHAR(255) NOT NULL UNIQUE,
```
255 символов ты ставишь, окей

20cognosce replied:
А не вылезем ли мы за лимит при генерации

20cognosce commented on src/main/java/faang/school/paymentservice/mapper/PendingOperationMapper.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    default String generateIdempotencyKey(PendingOperationDto dto) {
+        return IdempotencyKeyGenerator.generateKey(dto);
```
Не одобряю делать это в маппере
Лучше в сервисе, получили сущность к сохранению и тогда гегеним ей хэш
